In [1]:
import os
import cv2
import numpy as np
import matplotlib.pyplot as plt
import random 
import json

In [2]:
class Electronics:
    RAW_IMAGES = 'data'
    ROOT = "electronics"
    
    CAMERA = "camera"
    CELLPHONE = 'cellphone'
    CHARGER = 'charger'
    DESKTOP = 'desktop'
    DRONE = 'drone'
    HEADPHONE = 'headphone'
    LAPTOP = 'laptop'
    MOUSE = 'mouse'
    TELEVISION = 'television'
    REMOTE = 'remote'
    
    GRAY = 'gray'
    BGR = 'bgr'
    RGB = 'rgb'


> Class names of `electronics`

In [3]:
class_names = dict([(key, value) for (value, key ) in enumerate(os.listdir(Electronics.RAW_IMAGES))])
class_names

{'camera': 0,
 'cellphone': 1,
 'charger': 2,
 'desktop': 3,
 'drone': 4,
 'headphone': 5,
 'laptop': 6,
 'mouse': 7,
 'remote': 8,
 'television': 9}

> Lenght of all the image lists in the `dir`.

In [4]:
dir_count = np.array([len(os.listdir(os.path.join(Electronics.RAW_IMAGES, Electronics.CAMERA))),
len(os.listdir(os.path.join(Electronics.RAW_IMAGES, Electronics.CELLPHONE))),
len(os.listdir(os.path.join(Electronics.RAW_IMAGES, Electronics.CHARGER))),
len(os.listdir(os.path.join(Electronics.RAW_IMAGES, Electronics.DESKTOP))),
len(os.listdir(os.path.join(Electronics.RAW_IMAGES, Electronics.DRONE))),
len(os.listdir(os.path.join(Electronics.RAW_IMAGES, Electronics.HEADPHONE))),
len(os.listdir(os.path.join(Electronics.RAW_IMAGES, Electronics.LAPTOP))),
len(os.listdir(os.path.join(Electronics.RAW_IMAGES, Electronics.MOUSE))),
len(os.listdir(os.path.join(Electronics.RAW_IMAGES, Electronics.REMOTE))),
len(os.listdir(os.path.join(Electronics.RAW_IMAGES, Electronics.TELEVISION))),])
dir_count

array([260, 600, 260, 500, 240, 320, 600, 340, 300, 360])

> The lowest size of number of images in of a certain class is `240` which means our dataset
will contain `240` images per class.

In [5]:
max_len = np.min(dir_count)
max_len

240

In [6]:

    pass

In [7]:
def imageGenerator(catetory:str, size:int = max_len):
    images_list = os.listdir(os.path.join(Electronics.RAW_IMAGES, category))
    random.shuffle(images_list)
    for img_path in images_list[:size]:
        WIDTH, HEIGHT = 100, 100
        """
        Saving paths
        """
        image_path_gray = os.path.join(os.path.join(os.path.join(
            Electronics.ROOT, category), Electronics.GRAY))
        image_path_rgb = os.path.join(os.path.join(os.path.join(
            Electronics.ROOT, category), Electronics.RGB))
        image_path_bgr = os.path.join(os.path.join(os.path.join(
            Electronics.ROOT, category), Electronics.BGR))

        if os.path.exists(image_path_gray) == False:
            os.makedirs(image_path_gray)
        if os.path.exists(image_path_bgr) == False:
            os.makedirs(image_path_bgr)
        if os.path.exists(image_path_rgb) == False:
            os.makedirs(image_path_rgb)
        """
        Manipulating pixels
        """
        image_path = os.path.join(os.path.join(Electronics.RAW_IMAGES, category), img_path)
        image_bgr = cv2.resize(cv2.imread(image_path), (WIDTH, HEIGHT))
        image_gray = cv2.resize(cv2.imread(image_path, cv2.IMREAD_GRAYSCALE), (WIDTH, HEIGHT))
        image_rgb = cv2.resize(cv2.cvtColor(image_bgr, cv2.COLOR_BGR2RGB), (WIDTH, HEIGHT))

        """
        Save the manipulated images.
        """
        print(f"Saving {img_path}...")
        cv2.imwrite(f'{image_path_rgb}\\{img_path}' , image_rgb)
        cv2.imwrite(f'{image_path_bgr}\\{img_path}', image_bgr)
        cv2.imwrite(f'{image_path_gray}\\{img_path}', image_gray)
    print(f"\nSave History\nTotal:\t\t {max_len}\nCategory:\t {catetory.title()}\n\n")

> Generating ``images``.

In [8]:
categories = np.array([
    Electronics.CAMERA,
    Electronics.CELLPHONE,
    Electronics.CHARGER,
    Electronics.DESKTOP,
    Electronics.DRONE,
    Electronics.HEADPHONE,
    Electronics.LAPTOP,
    Electronics.MOUSE,
    Electronics.TELEVISION,
    Electronics.REMOTE,
])
categories

array(['camera', 'cellphone', 'charger', 'desktop', 'drone', 'headphone',
       'laptop', 'mouse', 'television', 'remote'], dtype='<U10')

In [9]:
for category in categories:
    imageGenerator(category)
with open(f'{Electronics.ROOT}\\class_names.json', 'w') as f:
    json.dump(class_names, f)
print("Saved")

Saving camera30.png...
Saving camera65.png...
Saving camera248.png...
Saving camera154.png...
Saving camera196.png...
Saving camera67.png...
Saving camera239.png...
Saving camera83.png...
Saving camera106.png...
Saving camera14.png...
Saving camera153.png...
Saving camera139.png...
Saving camera134.png...
Saving camera44.png...
Saving camera119.png...
Saving camera115.png...
Saving camera39.png...
Saving camera249.png...
Saving camera54.png...
Saving camera229.png...
Saving camera190.png...
Saving camera169.png...
Saving camera116.png...
Saving camera243.png...
Saving camera232.png...
Saving camera13.png...
Saving camera141.png...
Saving camera227.png...
Saving camera166.png...
Saving camera50.png...
Saving camera95.png...
Saving camera8.png...
Saving camera230.png...
Saving camera80.png...
Saving camera214.png...
Saving camera90.png...
Saving camera110.png...
Saving camera222.png...
Saving camera36.png...
Saving camera215.png...
Saving camera4.png...
Saving camera47.png...
Saving came

Saving cellphone282.png...
Saving cellphone583.png...
Saving cellphone21.png...
Saving cellphone305.png...
Saving cellphone202.png...
Saving cellphone208.png...
Saving cellphone392.png...
Saving cellphone330.png...
Saving cellphone252.png...
Saving cellphone161.png...
Saving cellphone391.png...
Saving cellphone460.png...
Saving cellphone408.png...
Saving cellphone34.png...
Saving cellphone142.png...
Saving cellphone233.png...
Saving cellphone228.png...
Saving cellphone532.png...
Saving cellphone52.png...
Saving cellphone353.png...
Saving cellphone145.png...
Saving cellphone88.png...
Saving cellphone494.png...
Saving cellphone279.png...
Saving cellphone66.png...
Saving cellphone50.png...
Saving cellphone49.png...
Saving cellphone301.png...
Saving cellphone571.png...
Saving cellphone318.png...
Saving cellphone300.png...
Saving cellphone404.png...
Saving cellphone122.png...
Saving cellphone587.png...
Saving cellphone131.png...
Saving cellphone22.png...
Saving cellphone359.png...
Saving ce

Saving charger92.png...
Saving charger214.png...
Saving charger178.png...
Saving charger57.png...
Saving charger90.png...
Saving charger183.png...
Saving charger84.png...
Saving charger233.png...
Saving charger179.png...
Saving charger250.png...
Saving charger244.png...
Saving charger96.png...
Saving charger28.png...
Saving charger228.png...
Saving charger88.png...
Saving charger55.png...
Saving charger187.png...
Saving charger173.png...
Saving charger82.png...
Saving charger75.png...
Saving charger152.png...
Saving charger72.png...
Saving charger156.png...
Saving charger105.png...
Saving charger258.png...
Saving charger39.png...
Saving charger61.png...
Saving charger36.png...
Saving charger35.png...
Saving charger3.png...
Saving charger235.png...
Saving charger211.png...
Saving charger256.png...
Saving charger124.png...
Saving charger107.png...
Saving charger26.png...
Saving charger62.png...
Saving charger47.png...
Saving charger155.png...
Saving charger222.png...
Saving charger207.pn

Saving drone155.png...
Saving drone191.png...
Saving drone2.png...
Saving drone229.png...
Saving drone78.png...
Saving drone188.png...
Saving drone231.png...
Saving drone18.png...
Saving drone172.png...
Saving drone101.png...
Saving drone103.png...
Saving drone52.png...
Saving drone226.png...
Saving drone197.png...
Saving drone96.png...
Saving drone129.png...
Saving drone235.png...
Saving drone238.png...
Saving drone186.png...
Saving drone59.png...
Saving drone89.png...
Saving drone54.png...
Saving drone141.png...
Saving drone104.png...
Saving drone239.png...
Saving drone108.png...
Saving drone124.png...
Saving drone208.png...
Saving drone90.png...
Saving drone20.png...
Saving drone112.png...
Saving drone237.png...
Saving drone46.png...
Saving drone114.png...
Saving drone110.png...
Saving drone131.png...
Saving drone150.png...
Saving drone210.png...
Saving drone196.png...
Saving drone149.png...
Saving drone107.png...
Saving drone99.png...
Saving drone40.png...
Saving drone135.png...
Sa

Saving headphone251.png...
Saving headphone39.png...
Saving headphone314.png...
Saving headphone225.png...
Saving headphone145.png...
Saving headphone247.png...
Saving headphone285.png...
Saving headphone320.png...
Saving headphone124.png...
Saving headphone319.png...
Saving headphone75.png...
Saving headphone172.png...
Saving headphone213.png...
Saving headphone284.png...
Saving headphone211.png...
Saving headphone13.png...
Saving headphone115.png...
Saving headphone57.png...
Saving headphone141.png...
Saving headphone71.png...
Saving headphone318.png...
Saving headphone15.png...
Saving headphone137.png...
Saving headphone107.png...
Saving headphone44.png...
Saving headphone195.png...
Saving headphone156.png...
Saving headphone16.png...
Saving headphone144.png...
Saving headphone276.png...
Saving headphone299.png...
Saving headphone224.png...
Saving headphone194.png...
Saving headphone290.png...
Saving headphone12.png...
Saving headphone264.png...
Saving headphone108.png...
Saving hea

Saving laptop286.png...
Saving laptop406.png...
Saving laptop349.png...
Saving laptop382.png...
Saving laptop124.png...
Saving laptop217.png...

Save History
Total:		 240
Category:	 Laptop


Saving mouse229.png...
Saving mouse224.png...
Saving mouse211.png...
Saving mouse209.png...
Saving mouse253.png...
Saving mouse110.png...
Saving mouse23.png...
Saving mouse54.png...
Saving mouse30.png...
Saving mouse295.png...
Saving mouse155.png...
Saving mouse297.png...
Saving mouse274.png...
Saving mouse141.png...
Saving mouse222.png...
Saving mouse315.png...
Saving mouse213.png...
Saving mouse94.png...
Saving mouse300.png...
Saving mouse4.png...
Saving mouse150.png...
Saving mouse105.png...
Saving mouse19.png...
Saving mouse234.png...
Saving mouse55.png...
Saving mouse37.png...
Saving mouse186.png...
Saving mouse9.png...
Saving mouse221.png...
Saving mouse108.png...
Saving mouse151.png...
Saving mouse197.png...
Saving mouse129.png...
Saving mouse191.png...
Saving mouse288.png...
Saving mouse269

Saving television56.png...
Saving television343.png...
Saving television240.png...
Saving television316.png...
Saving television203.png...
Saving television89.png...
Saving television202.png...
Saving television284.png...
Saving television264.png...
Saving television70.png...
Saving television356.png...
Saving television23.png...
Saving television68.png...
Saving television301.png...
Saving television164.png...
Saving television289.png...
Saving television237.png...
Saving television38.png...
Saving television299.png...
Saving television51.png...
Saving television145.png...
Saving television98.png...
Saving television220.png...
Saving television263.png...
Saving television173.png...
Saving television4.png...
Saving television168.png...
Saving television75.png...
Saving television32.png...
Saving television3.png...
Saving television123.png...
Saving television196.png...
Saving television39.png...
Saving television238.png...
Saving television337.png...
Saving television253.png...
Saving 

Saving remote52.png...
Saving remote236.png...
Saving remote270.png...
Saving remote242.png...
Saving remote175.png...
Saving remote259.png...
Saving remote131.png...
Saving remote18.png...
Saving remote246.png...
Saving remote66.png...
Saving remote96.png...
Saving remote284.png...
Saving remote46.png...
Saving remote178.png...
Saving remote77.png...
Saving remote152.png...
Saving remote102.png...
Saving remote297.png...
Saving remote123.png...
Saving remote202.png...
Saving remote208.png...
Saving remote125.png...
Saving remote117.png...
Saving remote118.png...
Saving remote103.png...
Saving remote112.png...
Saving remote101.png...
Saving remote252.png...
Saving remote216.png...
Saving remote145.png...
Saving remote227.png...
Saving remote180.png...
Saving remote261.png...
Saving remote122.png...
Saving remote298.png...
Saving remote111.png...
Saving remote15.png...
Saving remote21.png...
Saving remote150.png...
Saving remote174.png...
Saving remote106.png...
Saving remote113.png...


In [ ]:
> Done Proccesing images.